salt/grains/core.py

```diff
@@ -1390,7 +1396,8 @@ def os_data():
         shortname = distroname.replace(' ', '').lower()[:10]
         # this maps the long names from the /etc/DISTRO-release files to the
         # traditional short names that Salt has used.
-        grains['os'] = _OS_NAME_MAP.get(shortname, distroname)
+        if 'os' not in grains:
```

Comment:
@cachedout I don't see a problem here. The `'os'` key was always set from the keymap, but in our systems it is not always same. Hence @meaksh first trying to modify grain before picking standard key. In case not found anything, one from `_OS_NAME_MAP` is taken. I would say :+1: here, actually.

However, I see another trouble :boom: here: the whole code soon to be a better "winner" than a [Toyota design](http://www.safetyresearch.net/blog/articles/toyota-unintended-acceleration-and-big-bowl-%E2%80%9Cspaghetti%E2%80%9D-code). I mean, let's look at this monster function... It is horrid and needs to be splitted, refactored and cleaned up. IMHO. Yes, **only** in `develop`, no worries. :wink: 

Just my few cents.